pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -4355,38 +4356,50 @@ private PersistentReplicator getReplicatorReference(String replName, PersistentT
         }
     }
 
+    private CompletableFuture<Void> updatePartitionedTopicMetadataAsync(int numPartitions) {
+        CompletableFuture<Void> future = namespaceResources().getPartitionedTopicResources()
+                .updatePartitionedTopicAsync(topicName, p -> new PartitionedTopicMetadata(numPartitions));
+        future.exceptionally(ex -> {
+            // If the update operation fails, clean up the partitions that were created
+            getPartitionedTopicMetadataAsync(topicName, false, false).thenAccept(metadata -> {
+                int oldPartition = metadata.partitions;
+                for (int i = oldPartition; i < numPartitions; i++) {
+                    topicResources().deletePersistentTopicAsync(topicName.getPartition(i)).exceptionally(ex1 -> {
+                        log.warn("[{}] Failed to clean up managedLedger {}", clientAppId(), topicName,
+                                ex1.getCause());
+                        return null;
+                    });
+                }
```

Comment:
It looks like we provide a way that can shrink partitions. If the method used by other places might introduce unexpected behavior.

Reply:
No, it does not. The method only shrink the new partitions when `updatePartitionedTopicAsync` occur exception.